lib/Parse/ParseExpr.cpp

```diff
@@ -1003,6 +1003,48 @@ static bool canParseTypeOf(Parser &P) {
   return true;
 }
 
+static bool shouldWarnExcessiveRAngleInRange(SourceRange SR) {
```

Comment:
As long as we're going to be parsing them anyway, can you incorporate this check into `Parser::parseGenericArguments`?  You can also use a higher-level interface there by tracking the flow of tokens through the parser rather than re-lexing the buffer.